lib/puppet/pops/types/type_mismatch_describer.rb

```diff
@@ -598,7 +620,16 @@ def describe_struct_signature(params_struct, param_hash, missing_ok = false)
         value = param_hash[name]
         value_type = elem.value_type
         if param_hash.include?(name)
-          result << describe(value_type, TypeCalculator.singleton.infer_set(value), [ParameterPathElement.new(name)]) unless value_type.instance?(value)
+          if Puppet::Pops::Types::TypeFactory.deferred.implementation_class == value.class
+            if (df_return_type = get_deferred_function_return_type(value))
+              result << describe(value_type, df_return_type, [ParameterPathElement.new(name)]) unless matching_return_type?(df_return_type, value_type)
```

Comment:
I'm not sure what to do about parameterized classes. For example, if a class accepts a non-empty string `String[1]`, but the deferred function returns `String`, then the deferred function could return an empty string. Similarly, the class may accept `Pattern[/A-Z/]`. We could partially validate it by ensuring the [generalized]((https://github.com/puppetlabs/puppet/blob/c0f692088789217a35486714de2e9d53bfaf6622/lib/puppet/pops/types/types.rb#L185-L193) version of each type is assignable, like:

```ruby
value_type.generalize.assignable?(df_return_type.generalize)
```
But we won't actually know the length of string the deferred function returns until it's evaluated on the agent.

Reply:
It seems like this particular case is solved by the additional verification that can happen after the lazy-deference happens; is it reasonable to punt these sorts of things to some sort of client side deferred validation?

Reply:
For now I think we can only do a "best effort" validation on the server-side by comparing generalized types. The agent will also verify that the deferred function returns a value that matches its `return_type`. But the agent doesn't have access to the class parameter signatures, so it can't validate the value matches the signature more exactly, like the non-empty string case.

Reply:
I was thinking that a deferred function could be marked as having additional requirements for the return type, if the class parameter was more restrictive than the return_type here. But that seems out of scope for what this PR claims to do.